### PR TITLE
refactor!: replace yargs with Node.js util.parseArgs

### LINKED ISSUE #6122: ⚡️ Performance: Mocha 12 has larger install size despite fewer dependencies
### Performance Suggestion Checklist

- [x] I am using the latest version of Mocha.
- [x] I have read and understood the nuances around performance reports.
- [x] I have read and agree to Mocha's [Code of Conduct](https://github.com/mochajs/mocha/blob/main/.github/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md)
- [x] I have searched for [related issues](https://github.com/mochajs/mocha/issues?q=is%3Aissue) and [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20), but none matched my issue.
- [x] I want to provide a PR to resolve this

### Overview

Mocha 12 is a streamlining & modernization effort: we've removed some deps, transitioned the internals from CJS to ESM, etc. It's got fewer dependencies than 11 - which is good on its own!

However, the actual install size of `node_modules/` is larger: see https://packagephobia.com/result?p=mocha@12.0.0-beta-1 & https://packagephobia.com/result?p=mocha@12.0.0-rc.1. Which is an unfortunate regression. It comes from a few dependencies having gotten larger in their latest versions.

It'd be nice if we could also boast that Mocha 12 is _smaller_ than 11 in terms of install size. OTOH, I think it hurts Mocha 12's desirability to be _bigger_ than 11.

### Validations

```sh
for v in 11 12.0.0-rc.1; do
  rm -rf package* node_modules > /dev/null
  npm init --yes > /dev/null
  npm i mocha@$v > /dev/null
  echo "-- mocha@$v --"
  du -sk node_modules
done
```

```plaintext
-- mocha@11 --
11940   node_modules
-- mocha@12.0.0-rc.1 --
12996   node_modules
```

### Additional Info

I wish I'd gotten around to checking this before we got to the RC phase...

I suspect the easiest win for install size will be removing `yargs` altogether. This would generally be a good thing IMO, since the [Node.js-native `util.parseArgs`](https://nodejs.org/api/util.html#utilparseargsconfig) has been available since Node.js 16.17.0/18.3.0.

### PR DESCRIPTION
**Pull request by mark-wiemer**
### PR Checklist

- [x] Addresses an existing open issue: fixes #6122 (or is a [trivial change](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md#trivial-changes))
- [x] That issue was marked as [`status: accepting prs`](https://github.com/mochajs/mocha/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

Replaces `yargs` command dispatching with [Node.js-native `util.parseArgs`](https://nodejs.org/api/util.html#utilparseargsconfig). Marked as a breaking change out of an abundance of caution.

### Behavior Changes

- `--help` output is now generated by Mocha instead of yargs, so exact formatting, grouping, and wrapping may differ while preserving the same user-facing information.
- `--version` / `-V` and `--help` / `-h` are handled before run preparation, so they print version/help output even when other options such as `--require` point to modules that cannot be loaded.

---

(Ref #6128 and #6163 for paperwork issues)

### CODE REVIEW

File: lib/cli/run-option-metadata.cjs [hunk 1]
```diff
@@ -95,6 +95,67 @@ const aliases = {
...
+  config: "Path to config file",
+  delay: "Delay initial execution of root suite",
+  diff: "Show diff on failure",
+  "dry-run": "Report tests without executing them",
+  exit: "Force Mocha to quit after tests complete",
+  extension: "File extension(s) to load",
+  "fail-hook-affected-tests":
+    "Report tests as failed when affected by hook failures (before/beforeEach)",
+  "fail-zero": "Fail test run if no test(s) encountered",
+  "pass-on-failing-test-suite": "Not fail test run if tests were failed",
```
Comment (mark-wiemer) [inline lib/cli/run-option-metadata.cjs:114]: New PR: fix this yucky description

Comment (mark-wiemer) [reply lib/cli/run-option-metadata.cjs:114]: I'll open the new PR once we merge this PR, not gonna bother with a backing issue for this one, this comment thread is sufficient for me

File: lib/cli/run-option-metadata.cjs [hunk 2]
```diff
@@ -95,6 +95,67 @@ const aliases = {
...
+  fgrep: "Only run tests containing this string",
+  file: "Specify file(s) to be loaded prior to root suite execution",
+  "forbid-only": "Fail if exclusive test(s) encountered",
+  "forbid-pending": "Fail if pending test(s) encountered",
+  "full-trace": "Display full stack traces",
+  global: "List of allowed global variables",
```
Comment (mark-wiemer) [inline lib/cli/run-option-metadata.cjs:120]: Description of `global`: Another yucky help message, this should say "List allowed global variables", imperative

Comment (mark-wiemer) [reply lib/cli/run-option-metadata.cjs:120]: (Again, will be done in its own PR)

File: lib/cli/run.cjs [hunk 3]
```diff
@@ -21,359 +20,125 @@ const {
...
+      );
+    }
+
+    acc[pair[0]] = pair.length === 2 ? pair[1] : true;
+    return acc;
+  }, {});
+
+const normalizeRunOptions = (argv) => {
+  Object.keys(argv).forEach((name) => {
+    const reporterOptionMatch = name.match(/^reporter-options?\.(.+)$/);
```
Comment (mark-wiemer) [inline lib/cli/run.cjs:52]: Did we support `reporter-options` (plural style) before? If not, we should not add support for it now without consideration

```suggestion
    const reporterOptionMatch = name.match(/^reporter-option\.(.+)$/);
```

Comment (mark-wiemer) [reply lib/cli/run.cjs:52]: Ah, we had support in lib/cli/run-option-metadata.cjs:89 before, resolving this one. We also have a unit test for it, nice

File: lib/cli/run.cjs [hunk 4]
```diff
@@ -21,359 +20,125 @@ const {
...
 
-        if (argv.reporter === "markdown") {
-          throw createUnsupportedError(
-            "--reporter=markdown is mutually exclusive with --parallel",
-          );
-        }
+  if (argv.parallel) {
+    if (argv.file) {
+      throw createUnsupportedError(
+        "--parallel runs test files in a non-deterministic order, and is mutually exclusive with --file",
```
Comment (mark-wiemer) [inline lib/cli/run.cjs:107]: New PR: why mention "non-deterministic" in any `parallel` error message?

File: lib/cli/run.cjs [hunk 5]
```diff
@@ -21,359 +20,125 @@ const {
...
+        argv[normalizedName] = argv[name];
       }
+    }
+  });
+
+  if (Array.isArray(argv["reporter-option"])) {
+    argv["reporter-option"] = coerceReporterOption(argv["reporter-option"]);
+  }
+
+  return argv;
```
Comment (mark-wiemer) [inline lib/cli/run.cjs]: We should just return void here since we modify the argument directly

Comment (mark-wiemer) [inline lib/cli/run.cjs]: We should (implicitly) return void here since we modify the argument directly. Saves time wondering "oh, that returns something, does that mean it doesn't modify the argument?"

```suggestion
  }
```

File: lib/cli/run.cjs [hunk 6]
```diff
@@ -21,359 +20,125 @@ const {
...
-          );
-        }
+const validateRunOptions = (argv) => {
+  // "one-and-dones"; help and version are handled by the top-level CLI.
+  const oneAndDoneOptions = Object.keys(ONE_AND_DONES).filter((opt) =>
+    Boolean(argv[opt]),
+  );
+  if (oneAndDoneOptions.length > 1) {
+    throw createUnsupportedError(
+      `Arguments ${oneAndDoneOptions.join(" and ")} are mutually exclusive`,
```
Comment (mark-wiemer) [inline lib/cli/run.cjs:79]: Hm, just want to make sure we have comprehensive tests here

File: lib/cli/run.cjs [hunk 7]
```diff
@@ -21,359 +20,125 @@ const {
...
-      if (argv.parallel) {
-        // yargs.conflicts() can't deal with `--file foo.js --no-parallel`, either
-        if (argv.file) {
-          throw createUnsupportedError(
-            "--parallel runs test files in a non-deterministic order, and is mutually exclusive with --file",
-          );
-        }
+const validateRunOptions = (argv) => {
+  // "one-and-dones"; help and version are handled by the top-level CLI.
+  const oneAndDoneOptions = Object.keys(ONE_AND_DONES).filter((opt) =>
```
Comment (mark-wiemer) [inline lib/cli/run.cjs:74]: Should we use ONE_AND_DONE_ARGS here?


```suggestion
  const oneAndDoneOptions = Object.keys(ONE_AND_DONE_ARGS).filter((opt) =>
```

Comment (mark-wiemer) [reply lib/cli/run.cjs:74]: I know we have the special `h` and `v` handling above, but maybe we can remove that instead?

File: lib/cli/run.cjs (hunk 6 above)
Comment (mark-wiemer) [reply lib/cli/run.cjs:79]: added tests

File: lib/cli/run.cjs (hunk 7 above)
Comment (mark-wiemer) [reply lib/cli/run.cjs:74]: Ah, ONE_AND_DONES have functions, ONE_AND_DONE_ARGS don't. ONE_AND_DONE_ARGS is used for short-circuiting even earlier, I'm fine to keep things as they are, don't want to rock the boat too much, esp in an already-large PR :)

File: lib/cli/run.cjs (hunk 3 above)
Comment (mark-wiemer) [reply lib/cli/run.cjs:52]: We did support it :)

File: lib/cli/run.cjs (hunk 4 above)
Comment (mark-wiemer) [reply lib/cli/run.cjs:107]: https://github.com/mochajs/mocha/issues/6165

File: lib/cli/run-option-metadata.cjs (hunk 2 above)
Comment (mark-wiemer) [reply lib/cli/run-option-metadata.cjs:120]: https://github.com/mochajs/mocha/issues/6165

File: lib/cli/run-option-metadata.cjs (hunk 1 above)
Comment (mark-wiemer) [reply lib/cli/run-option-metadata.cjs:114]: https://github.com/mochajs/mocha/issues/6165

**Comment by mark-wiemer** [comment]
(Converting back to draft while I handle some local AI-generated feedback)

**Comment by mark-wiemer** [comment]
(It was all hallucinations)

**Review by JoshuaKGoldberg — APPROVED** [review]
Swell! We've both worked on this and it looks very good to me. I'm still a little nervous about edge cases in the CLI, but:

* This area has a lot of test coverage, including some added points (IMO the coverage drop is inaccurate and misleading)
* It's a major version that's going into RC anyway, little oddities are expected (and always fixable after the fact)

Merging now. 🚀
